### ISSUE TITLE
Range of sin(x) and cos(x)

### ISSUE THREAD

**Issue by k-arrows** [llvm/llvm-project#68301]
As you know, the absolute values of sin(x) and cos(x) never exceed 1. GCC trunk seems to be able to handle this.
https://godbolt.org/z/fMTzTd993

**Comment by dtcxzyw** [comment]
I don't think this optimization is meaningful for real-world applications. @nikic Do we have a plan to track the range of fp numbers like `ConstantRange`?

**Comment by k-arrows** [comment]
> I don't think this optimization is meaningful for real-world applications.

I agree with you on this point. But I don't think it's unnatural to do such a check of a function with a fixed value range.

**Comment by dtcxzyw** [comment]
> > I don't think this optimization is meaningful for real-world applications.
> 
> I agree with you on this point. But I don't think it's unnatural to do such a check of a function with a fixed value range.

I do favor generalizing this by tracking the range of fp numbers as `ConstantRange` does.

**Comment by dtcxzyw** [comment]
See also: https://discourse.llvm.org/t/floating-point-range-checks/34923
